docs/specifications/tests/Zone-TP/zone11.md

```diff
@@ -0,0 +1,208 @@
+       3. Add a pair consisting of the *Name Server IP* and the lowercase
+          string thus derived from the RDATA field to the *SPF-Policies* set.
+
+   6. Go to the next name server.
+
+5. If the *SPF-Policies* set is empty, then output
+   *[Z11_UNABLE_TO_CHECK_FOR_SPF]* and terminate the test.
+
+6. If all the pairs in the *SPF-Policies* set contain empty strings, then
+   output *[Z11_NO_SPF_FOUND]* and terminate the test.
+
+7. Compare the set of *SPF-Policies* retrieved from all name servers. If at
+   least two different name servers have returned different sets of SPF
+   policies, then:
+
+   1. Output *[Z11_INCONSISTENT_SPF_POLICIES]*.
+   2. Group *SPF-Policies* by equal sets of SPF policies, such that a set of
+      SPF policies is mapped to the list of *Name Server IPs* that returned it.
+   3. For each such group of name servers, output
+      *[Z11_DIFFERENT_SPF_POLICIES_FOUND]*.
+   4. Terminate the test.
+
+8. If the *SPF-Policies* set contains at least two pairs with the same IP
+   address, then output *[Z11_SPF1_MULTIPLE_RECORDS]* with the list of
+   nameservers that returned more than one SPF policy and terminate the test.
+
+9. The following steps assume that all pairs in the *SPF-Policies* set have
+   the same string ("SPF policy").
+
+10. If the *SPF Policy* does not [pass the syntax check][passing the syntax
+    check] for SPF version 1 records, then output *[Z11_SPF1_SYNTAX_ERROR]* and
+    terminate the test.
+
+11. If no other message was outputted by this test case, then output
+    *[Z11_SPF1_SYNTAX_OK]*.
+
+## Outcome(s)
+
+The outcome of this Test Case is "fail" if there is at least one message
+with the severity level *[ERROR]* or *[CRITICAL]*.
+
+The outcome of this Test Case is "warning" if there is at least one message
+with the severity level *[WARNING]*, but no message with severity level
+*ERROR* or *CRITICAL*.
+
+In other cases, no message or only messages with severity level
+*[INFO]* or *[NOTICE]*, the outcome of this Test Case is "pass".
+
+## Special procedural requirements
+
+If either IPv4 or IPv6 transport is disabled, skip sending queries over that
+transport protocol. A message will be outputted reporting that the transport
+protocol has been skipped.
+
+## Intercase dependencies
+
+None.
+
+## Terminology
+
+* "SPF TXT record" - The term is used to refer to a TXT resource record which,
+  after [concatenating][concatenate] all strings within that resource record
+  into one string, yields a string either equal to `v=spf1` or starting with
+  `v=spf1` followed by a space, irrespective of character case.
+
+* "concatenate" - The term is used to refer to the conversion of a TXT
+  resource record’s data to a single contiguous string, as specified in [RFC
+  7208, section 3.3][RFC 7208#3.3].
+
+* "passing the syntax check" - The term is used in this document to refer to
+  text that is valid according to the ABNF grammar published in [RFC 7208]
+  starting from [section 4.5][RFC 7208#4.5].
```

Comment:
It is not easy to many verify this unless you are familiar with the syntax. Are there any on-line SPF validation tools that we could refer to, not as an normative reference, but something that the reader could use to check with?

(I realize that the same comment could be put on the DNSSEC test case specification where validation is done.)

Reply:
There is a rather nice SPF syntax checker here: <https://vamsoft.com/support/tools/spf-syntax-validator>. It’s one of the few where you can actually input an SPF policy to check, instead of a domain name.

In my draft implementation for which I’ve yet to create a pull request, I am using the syntax check from the Mail::SPF module. Based on a sample of 50 SPF policies with Mail::SPF deems to be in error, the aforementioned syntax checker says that 40 have a syntax error, 9 are “OK with warnings” (for example `v=spf1 ip4:192.0.2.25 +a +mx +include:spf.jabatus.fr ~allgoogle-site-verification=yjZfTeZ8FHjaH4tY8UZvrEc_enSwtJFRtyJtbNy7RPg`, with the actual IP address replaced with one in a documentation range) and only 1 was “OK”.

The only discrepancy I was with an SPF policy containing a newline (U+000A) character. In that case, Mail::SPF will raise an error, but the online tool said it was OK. However, the ABNF grammar for SPF only allows spaces (U+0020) as whitespace, so Mail::SPF is correct here.